x-pack/plugins/lens/public/react_embeddable/expressions/expression_params.ts

```diff
@@ -0,0 +1,231 @@
+    if (!uiActions?.getTriggerCompatibleActions) {
+      return false;
+    }
+    if (
+      isLensTableRowContextMenuClickEvent(event) ||
+      isLensMultiFilterEvent(event) ||
+      isLensFilterEvent(event)
+    ) {
+      const actions = await uiActions.getTriggerCompatibleActions(
+        VIS_EVENT_TO_TRIGGER[event.name],
+        {
+          data: event.data,
+          embeddable: api,
+        }
+      );
+
+      return actions.length > 0;
+    }
+
+    return false;
+  };
+}
+
+function buildGetCompatibleCellValueActions(
+  api: LensApi,
+  { uiActions }: LensEmbeddableStartServices
+): GetCompatibleCellValueActions {
+  return async (data) => {
+    if (!uiActions?.getTriggerCompatibleActions) {
+      return [];
+    }
+    const actions: Array<Action<CellValueContext>> = (await uiActions.getTriggerCompatibleActions(
+      CELL_VALUE_TRIGGER,
+      { data, embeddable: api }
+    )) as Array<Action<CellValueContext>>;
+    return actions
+      .sort((a, b) => (a.order ?? Infinity) - (b.order ?? Infinity))
+      .map((action) => ({
+        id: action.id,
+        type: action.type,
+        iconType: action.getIconType({ embeddable: api, data, trigger: cellValueTrigger })!,
+        displayName: action.getDisplayName({ embeddable: api, data, trigger: cellValueTrigger }),
+        execute: (cellData) =>
+          action.execute({ embeddable: api, data: cellData, trigger: cellValueTrigger }),
+      }));
+  };
+}
+
+export async function getExpressionRendererParams(
+  state: LensRuntimeState,
+  {
+    unifiedSearch,
+    settings: { syncColors = true, syncCursor = true, syncTooltips = false },
+    services,
+    disableTriggers = false,
+    getExecutionContext,
+    searchSessionId,
+    abortController,
+    onRender,
+    handleEvent,
+    onData = noop,
+    logError,
+    api,
+    addUserMessages,
+    updateBlockingErrors,
+  }: GetExpressionRendererPropsParams
+): Promise<{
+  params: ExpressionWrapperProps | null;
+  abortController?: AbortController;
+  indexPatterns: IndexPatternMap;
+  indexPatternRefs: IndexPatternRef[];
+  activeVisualizationState?: unknown;
+  activeDatasourceState?: unknown;
+}> {
+  const { expressionRenderer, documentToExpression, data, injectFilterReferences } = services;
+
+  const searchContext = getMergedSearchContext(state, unifiedSearch, {
+    data,
+    injectFilterReferences,
+  });
+
+  const {
+    expression,
+    indexPatterns,
+    indexPatternRefs,
+    activeVisualizationState,
+    activeDatasourceState,
+  } = await getExpressionFromDocument(state.attributes, documentToExpression);
+
+  // if at least one indexPattern is time based, then the Lens embeddable requires the timeRange prop
+  // this is necessary for the dataview embeddable but not the ES|QL one
+  if (
+    isSearchContextIncompatibleWithDataViews(api, searchContext, indexPatternRefs, indexPatterns)
+  ) {
+    addUserMessages([getSearchContextIncompatibleMessage()]);
+  }
```

Comment:
This checks now triggers a different behavior in Canvas causing the embeddable to stop rendering if you remove the default configured Custom Time Range.
I can reproduce it by adding the `[Logs] bytes distibuton` chart from the library to an empty canvas.
The same happen also if you add a time filter element and if you add the `filters |` function before the embeddable expression configuration.
I've checked in `main` and if you remove the custom time range the embeddable continues to render correctly.

<img width="540" alt="Screenshot 2024-11-06 at 17 58 28" src="https://github.com/user-attachments/assets/42b2bb08-37b0-4fb5-838e-2eec2d584f77">

Reply:
I had a go to debug this specific scenario with Canvas on both legacy and new Embeddable system.
This is what is happening in there:
* with the time filter set at panel level Canvas will just forward this panel setting downstream to the Embeddable, all works ✅ 
* when the panel time filter is removed the panel badge is removed BUT the chart fails to render in both cases ❌ 
  * on the new embeddable (this PR) this specific validation is executed at each render
  * on the legacy embeddable this specific validation is not executed at each render, only at bootstrap. So the renderer won't update the panel other than removing the badge.
     * even when adding a `filters` or `timefilter` the timeRange is not forwarding

I've tested this in `main`:
![canvas_timefilter_broken](https://github.com/user-attachments/assets/eea2d855-62eb-46ff-9d8c-8f55dee4714b)

The chart is preserving the original timeRange ignoring any piped filter (as Canvas is not passing anything to the Embeddable). So it broke in `main` for a while and due to this validation bug it wasn't spotted until now.

In short:
* I could add some specific `Canvas` check in the validation logic to reproduce the legacy embeddable behaviour
* Keep the current validation check on each render to other applications

Reply:
As discussed offline I've pushed a fix only for Canvas here:  2f246fe

Reply:
I've created an issue to track this https://github.com/elastic/kibana/issues/199603